R/weighted_interval_score.R

```diff
@@ -0,0 +1,73 @@
+#' Compute weighted interval score
+#'
+#' Weighted interval score (WIS), a well-known quantile-based
+#' approximation of the commonly-used continuous ranked probability score
+#' (CRPS). WIS is a proper score, and can be thought of as a distributional
+#' generalization of absolute error. For example, see [Bracher et
+#' al. (2020)](https://arxiv.org/abs/2005.12881) for discussion in the context
+#' of COVID-19 forecasting.
+#'
+#' @param x dist_quantiles. A vector of class [dist_quantiles()].
+#' @param actual double. Actual value(s)
+#' @param ... not used
+#'
+#' @return a vector of nonnegative scores.
+#'
+#' @export
+#' @examples
+#' quantile_levels <- c(.2, .4, .6, .8)
+#' predq_1 <- 1:4 #
+#' predq_2 <- 8:11
+#' dstn <- dist_quantiles(list(predq_1, predq_2), quantile_levels)
+#' actual <- c(3.3, 7.1)
+#' weighted_interval_score(dstn, actual)
+#'
+#' # Using some actual forecasts --------
+#' library(dplyr)
+#' jhu <- case_death_rate_subset %>%
+#'   filter(time_value >= "2021-10-01", time_value <= "2021-12-01")
+#' preds <- flatline_forecaster(
+#'   jhu, "death_rate",
+#'   flatline_args_list(quantile_levels = c(.01, .025, 1:19 / 20, .975, .99))
+#' )$predictions
+#' actuals <- case_death_rate_subset %>%
+#'   filter(time_value == as.Date("2021-12-01") + 7) %>%
+#'   select(geo_value, time_value, actual = death_rate)
+#' preds <- left_join(preds, actuals,
+#'   by = c("target_date" = "time_value", "geo_value")
+#' ) %>%
+#'   mutate(wis = weighted_interval_score(.pred_distn, actual))
+#' preds
+weighted_interval_score <- function(x, actual, ...) {
+  UseMethod("weighted_interval_score")
+}
+
+#' @export
+weighted_interval_score.default <- function(x, actual, ...) {
+  cli_abort("Weighted interval score can only be calculated if `x` has class {.cls distribution}.")
```

Comment:
A `distribution` is a vector container that can contain 1 or more `dist_quantiles`. But it can contain different types of distributions simultaneously. Like:
``` r
library(distributional)
dstn <- c(dist_normal(), dist_quantiles(1:4, 1:4/5))
```